### PR TITLE
run in docker mode + system tests

### PR DESCRIPTION
**Pull request by sanderegg**
This PR brings:
- running the dask-gateway in a docker container is now stable:
  - each running docker container is connected to the so-called docker_gw_bridge network on the host (typically accessible through IPs like 172.18.0.X) --> this is why integration tests were running fine, but when containerizing the dask-gateway it would fail. This is now fixed.
  - added a system test to check that this is always running fine

### CODE REVIEW

File: tests/test_deploy.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+            assert "workers" in cluster.scheduler_info
+            assert len(cluster.scheduler_info["workers"]) == _NUM_WORKERS
+            workers = deepcopy(cluster.scheduler_info["workers"])
+            print(
+                f"!-- {_NUM_WORKERS} are running,"
+                f" [{json.dumps(attempt.retry_state.retry_object.statistics)}]",
+            )
+
+    # now check all this is stable
+    _SECONDS_STABLE = 6
```
Comment (pcrespov) [inline tests/test_deploy.py:173]: i do not understand the notation here?  you check six times that the info on workers fit ... why SECONDS? what does stability has to do there as well?

File: tests/test_deploy.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+@pytest.fixture
+async def dask_gateway_stack_deployed_services(
+    osparc_gateway_server_root_dir: Path,
+    loop: asyncio.AbstractEventLoop,
+    docker_swarm,
+    aiohttp_client: aiohttp.ClientSession,
+    dask_gateway_entrypoint: str,
+):
+    print("--> Deploying osparc-dask-gateway stack...")
+    process = await asyncio.create_subprocess_exec(
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply tests/test_deploy.py:92]: no I like being fully async

File: tests/test_deploy.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+
+
+@pytest.fixture(scope="session")
+def dask_gateway_entrypoint() -> str:
+    return f"http://{_get_this_computer_ip()}:8000"
+
+
+@pytest.fixture(scope="session")
+def dask_gateway_password() -> str:
+    return "asdf"
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply tests/test_deploy.py:80]: hmm... for testing locally it is pretty annoying to use the fake for this at the moment. but for later yes

File: tests/test_deploy.py (hunk 1 above)
Comment (sanderegg) [reply tests/test_deploy.py:173]: you are right... was not thinking straight. it needs to sleep a bit in between.

File: services/osparc-gateway-server/src/osparc_gateway_server/backend/utils.py [hunk 4]
```diff
@@ -283,3 +284,31 @@ async def create_docker_secrets_from_tls_certs_for_cluster(
...
+        "--port": f"{OSPARC_SCHEDULER_PORT}",
+        "--host": scheduler_service_name,
+    }
+
+
+def modify_cmd_argument(
+    cmd: List[str], argument_name: str, argument_value: str
+) -> List[str]:
+    modified_cmd = deepcopy(cmd)
+    try:
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply services/osparc-gateway-server/src/osparc_gateway_server/backend/utils.py:309]: yes

File: tests/Makefile [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+install-dev install-prod install-ci: _check_venv_active ## install requirements in dev/prod/ci mode
+	# installing in $(subst install-,,$@) mode
+	pip-sync requirements/$(subst install-,,$@).txt
+
+
+.PHONY: tests
+tests: _check_venv_active ## runs all tests [DEV]
+	# running unit tests
+	pytest -vv \
+			--color=yes \
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply tests/Makefile:25]: hmm not too sure why this looks like this here... it seems like I have only tabs...

File: services/osparc-gateway-server/src/osparc_gateway_server/backend/utils.py [hunk 6]
```diff
@@ -283,3 +284,31 @@ async def create_docker_secrets_from_tls_certs_for_cluster(
             secret_data=cluster.tls_key.decode(),
         ),
     ]
+
+
+OSPARC_SCHEDULER_PORT: Final[int] = 8786
```
Comment (GitHK) [inline services/osparc-gateway-server/src/osparc_gateway_server/backend/utils.py:289]: minor: maybe have this at the start of the file

File: services/osparc-gateway-server/src/osparc_gateway_server/backend/utils.py [hunk 7]
```diff
@@ -283,3 +284,31 @@ async def create_docker_secrets_from_tls_certs_for_cluster(
...
+
+
+def get_osparc_scheduler_cmd_modifications(
+    scheduler_service_name: str,
+) -> Dict[str, str]:
+    # NOTE: the healthcheck of itisfoundation/dask-sidecar expects the dashboard
+    # to be on port 8787
+    # (see https://github.com/ITISFoundation/osparc-simcore/blob/f3d98dccdae665d23701b0db4ee917364a0fbd99/services/dask-sidecar/Dockerfile)
+    return {
+        "--dashboard-address": ":8787",
```
Comment (GitHK) [inline services/osparc-gateway-server/src/osparc_gateway_server/backend/utils.py:299]: minor: am I mistaken to suggest this?
```suggestion
        "--dashboard-address": f":{OSPARC_SCHEDULER_PORT}",
```

File: services/osparc-gateway-server/tests/integration/conftest.py [hunk 8]
```diff
@@ -70,9 +71,30 @@ def _convert_to_dict(c: Union[traitlets.config.Config, Dict]) -> Dict[str, Any]:
...
+    If the dask-gateway-server is running in the host then:
+    - dask-scheduler must start with "" for --host, so the dask-scheduler defines its IP as being in docker_gw_bridge (172.18.0.X), accessible from the host
+    When the dask-gateway-server is running as a docker container, then the --host must be set
+    as "cluster_X_scheduler" since this is the hostname of the container and resolves into the dask-gateway network
+    """
+    mocker.patch(
+        "osparc_gateway_server.backend.osparc.get_osparc_scheduler_cmd_modifications",
+        autospec=True,
+        return_value={
+            "--dashboard-address": ":8787",
```
Comment (GitHK) [inline services/osparc-gateway-server/tests/integration/conftest.py:86]: same question as above here

File: .github/workflows/gateway.yml [hunk 9]
```diff
@@ -110,7 +110,48 @@ jobs:
         with:
           flags: integrationtests #optional
 
-  build:
+  system-tests:
+    timeout-minutes: 10
+    strategy:
+      matrix:
+        python: [3.9]
```
Comment (GitHK) [inline .github/workflows/gateway.yml:117]: no 3.8 version?

**Review by GitHK — APPROVED** [review]
👍 some minor remarks

File: services/osparc-gateway-server/src/osparc_gateway_server/backend/utils.py (hunk 7 above)
Comment (sanderegg) [reply services/osparc-gateway-server/src/osparc_gateway_server/backend/utils.py:299]: the suggestion is wrong, but I will create a variable.
this port is the scheduler dashboard port

File: services/osparc-gateway-server/tests/integration/conftest.py (hunk 8 above)
Comment (sanderegg) [reply services/osparc-gateway-server/tests/integration/conftest.py:86]: same answer

File: .github/workflows/gateway.yml (hunk 9 above)
Comment (sanderegg) [reply .github/workflows/gateway.yml:117]: no. this is due to the fact that piwheels does only provide pre-built libraries for python 3.7 and 3.9. so I went for the more recent.

**Pull request merged**